cborg/tests/Tests/Properties.hs

```diff
@@ -380,6 +382,38 @@ prop_decodeRefdecodeImp _ x =
     t   = Proxy :: Proxy t
 
 
+-- | The property corresponding to the following part of the commuting diagram.
+--
+-- This is a round trip property, with the production implementation of the
+-- encoder and decoder, involving toFlatTerm and fromFlatTerm.
+--
+-- >         canon
+-- >    Ref -------▷Ref            ▶ FlatTerm ▶  
+-- >                 |           ╱            │
+-- >                 |          ╱toFlatTerm   │
+-- >                 |         ╱              │dec
+-- >          fromRef|    ▶Enc▶               │fromFlatTerm
+-- >                 |   /                    │
+-- >                 |  /enc                  │
+-- >                 ▽ /                      ▼
+-- >                Imp ─────────────────────▶Imp
+-- >                              id
+--
+-- > (fromFlatTerm dec_imp . toFlatTerm . enc_imp) imp = Right imp
+--
+prop_toFromFlatTerm :: forall t. Token t => Proxy t -> t -> Bool
+prop_toFromFlatTerm _ x =    
+  
+    liftA2 eq (fn enc) (Right imp) == Right True
```

Comment:
This could be expressed a bit more elegantly if you followed the same trick as `deserialiseImp` which is basically a wrapper that throws an exception of the result is a decode failure.

Here, you'd make a local wrapper around `fromFlatTerm` that just throws an exception in the `Left` case. Something like
```
  where
    deserialiseFromFlatTerm dec =
      case fromFlatTerm dec flatterm of
        Left _ -> error "fromFlatTerm: decode failure"
        Right x -> x
```
Then your overall property can match the diagram directly (and like in your comment).
```
(deserialiseFromFlatTerm (decodeImp t) . toFlatTerm . encodeImp t) imp  `eq`  imp
```
The idea with these commuting diagrams is that the property can be read off directly by composing all the arrows. So this reads most nicely when the property is exactly that.

Reply:
Yes, correct Implemented accordingly.